New API, refactored specs

andrewvc opened the pull request:
This is a WIP for a refactor that fixes or partially fixes a number of open issues:

https://github.com/logstash-plugins/logstash-input-http_poller/issues/7
https://github.com/logstash-plugins/logstash-input-http_poller/issues/9
https://github.com/logstash-plugins/logstash-input-http_poller/issues/6
https://github.com/logstash-plugins/logstash-input-http_poller/issues/3
https://github.com/logstash-plugins/logstash-input-http_poller/issues/2

Specs do not yet pass correctly

maximede commented:
I think it would make sense if all the failing requests added a "_http_request_failure" tag on top of the _http_request_failure object. Also, I have the feeling that it would be more consistent with the "_jsonparsefailure" tag

andrewvc commented:
You're totally right, that should be a tag not a field

On Monday, June 29, 2015, Maxime Deravet <notifications@github.com> wrote:

> I think it would make sense if all the failing requests added a
> "_http_request_failure" tag on top of the _http_request_failure object.
> Also, I have the feeling that it would be more consistent with the
> "_jsonparsefailure" tag
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/logstash-plugins/logstash-input-http_poller/pull/10#issuecomment-116879256>
> .
>

maximede commented:
Well, I think that adding the object also make sense, getting the error message in the event would be a huge plus to track down a possible problem

andrewvc commented:
Mm that's right I forgot that I put that metadata there for a reason. I'll
just add the tag to the pr

On Monday, June 29, 2015, Maxime Deravet <notifications@github.com> wrote:

> Well, I think that adding the object also make sense, getting the error
> message in the event would be a huge plus to track down a possible problem
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/logstash-plugins/logstash-input-http_poller/pull/10#issuecomment-116879591>
> .
>

maximede commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -91,77 +51,73 @@ def run(queue)
-      end
-    end
+    event["_http_request_failure"] = {
+      "url" => url,
```
I'm wondering if we are not duplicating information here. `url` and `name` are already in the main event.

andrewvc replied:
The idea here was that `@metadata` is not persisted by default. If you're getting errors it could be frustrating to find out that the info is being lost.

maximede replied:
Ok, understood. Makes sense then

andrewvc replied:
It could probably use an explanatory comment however

On Tuesday, June 30, 2015, Maxime Deravet <notifications@github.com> wrote:

> In lib/logstash/inputs/http_poller.rb
> <https://github.com/logstash-plugins/logstash-input-http_poller/pull/10#discussion_r33595854>
> :
>
> >
> > -      if (@truststore_password)
> > -        c.merge!(truststore_password: @trust_store_password)
> > -      end
> > -    end
> > +    event["_http_request_failure"] = {
> > +      "url" => url,
>
> Ok, understood. Makes sense then
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/logstash-plugins/logstash-input-http_poller/pull/10/files#r33595854>
> .
>

jordansissel commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -91,77 +51,73 @@ def run(queue)
+    event["@metadata"] = event_metadata(name, url, response)
+    queue << event
+  rescue StandardError, java.lang.Exception => e
+    @logger.error? && @logger.error("Error eventifying response!", exception: e, name: name, url: url, response: response)
```
We're still in the midst of style discussions, can we use the older hash-rocket syntax here instead until that's resolved?

andrewvc replied:
can do

jordansissel commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -91,77 +51,73 @@ def run(queue)
-    c
+    queue << event
+  rescue StandardError, java.lang.Exception => e
+      @logger.error("Cannot read URL! Also, couldn't send the error as an event!(#{exception}/#{exception.message})", :name => name, :url => url)
```
Try to avoid string interpolation in logger.error calls, move those to the context hash:

```@logger.error("Some plain text message", :exception => exception, :exception_message => exception.message, ...)``` or similar.

andrewvc replied:
Makes sense.

jordansissel commented on spec/inputs/http_poller_spec.rb:
```diff
@@ -2,87 +2,151 @@
+  describe "#run" do
+    it "should run at the specified interval" do
+      expect(Stud).to receive(:interval).with(default_interval).once
+      subject.send(:run, double("queue"))
```
Curious - Why use `Object#send` here if the method name isn't dynamic - specifically, why not subject.run(double("queue")) ?

andrewvc replied:
For `Object#send` the main reason is that it's a private method, and that's the only way I know of to call it.  Is there a better way in rspec?

jordansissel replied:
ooh, good call. Let's add a comment there to note that you're using send to call the private method (lol private, ruby!)

jordansissel commented on spec/inputs/http_poller_spec.rb:
```diff
@@ -2,87 +2,151 @@
+
+  shared_examples "unprocessable_requests" do
+    let(:poller) {
+      p = LogStash::Inputs::HTTP_Poller.new(settings)
```
Should we move the `allow(p)...` to a `before { }` block?

andrewvc replied:
yes! Good call.

jordansissel commented on spec/inputs/http_poller_spec.rb:
```diff
@@ -2,87 +2,151 @@
+      p
+    }
+    subject(:event) {
+      poller.send(:run_once, queue)
```
similar to my `allow` comment; it feels weird having let/subject blocks _do_ things, rather than declare things. I admit in some cases it's necessary, but my preference is to lean towards having before blocks _do_ things and let/subject blocks simply declare them

jordansissel commented on spec/inputs/http_poller_spec.rb:
```diff
@@ -2,87 +2,151 @@
+    }
+
+    before do
+      event # materialize the subject
```
Is this necessary? Invoking 'event' like the next line `expect(event)` will naturally materialize things anyway

jordansissel commented on spec/inputs/http_poller_spec.rb:
```diff
@@ -2,87 +2,151 @@
   end
 
-  context "with json codec, when remote server responds" do
+  describe "with a non responsive server" do
```
my preference for prepositional phrases in rspec is to use the 'context' word instead of 'describe'

`context "with a non responsive server" do`

etc

Execution-wise, they are equivalent (describe and context are aliases of the same method in RSpec), but I feel the `context` word has better meaning given positional information (with ...) is context

jordansissel commented on spec/inputs/http_poller_spec.rb:
```diff
@@ -2,87 +2,151 @@
-      expect(event["success"]).to eq(true)
+    describe "due to a bogus port number" do
+      let(:name) { default_name }
+      let(:url) { "http://127.0.0.1:9999999" }
```
we can add randomness here, perhaps?

```ruby
let(:invalid_port) { Flores::Random.integer(65536..1000000) }
let(:url) { "http://127.0.0.1:#{invalid_port}" }
```

jordansissel commented on spec/inputs/http_poller_spec.rb:
```diff
@@ -2,87 +2,151 @@
+  end
+
+  describe "a codec mismatch" do
+    let(:inst) {
```
inst?

andrewvc replied:
Shorthand for instance, I'll make it long form

jordansissel replied:
Sounds like the subject of a test, any thoughts on using subject instead?

andrewvc replied:
even better idea

andrewvc replied:
Actually the message is the subject, will be adjusting that

jordansissel commented on spec/inputs/http_poller_spec.rb:
```diff
@@ -2,87 +2,151 @@
-      expect{ logstash_queue.pop}.not_to raise_error
-      event = logstash_queue.pop
+    let(:qmsg) {
+      inst.client.stub(default_url, body: LogStash::Json.dump(payload), code: 202)
```
hashrocket syntax here until https://github.com/elastic/logstash/pull/3485 is resolved

jordansissel commented on spec/inputs/http_poller_spec.rb:
```diff
@@ -2,87 +2,151 @@
-      subject = LogStash::Inputs::HTTP_Poller.new(settings.merge({"urls" => {"url1" => "http://mytesturl.mydomain"}}))
-      subject.client.stub("http://mytesturl.mydomain", body:"not found",  code: 404)
+  shared_examples("matching metadata") {
+    let(:metadata) { event["@metadata"]["http_poller"] }
```
we should use fieldref syntax here:

```event["[@metadata][http_poller]"]``` even though it looks weird.

jordansissel commented on spec/inputs/http_poller_spec.rb:
```diff
@@ -2,87 +2,151 @@
+      "codec" => "json"
+    }
+  }
+  let(:mod) { LogStash::Inputs::HTTP_Poller }
```
mod? or klass? plugin_class? plugin_type?

andrewvc replied:
Ah yeah, it should be `klass`

jordansissel commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -22,6 +22,7 @@
 #    }
```
Let's also document the `@metadata` that is produced by this plugin

jordansissel commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -91,77 +51,73 @@ def run(queue)
+
+    # Some exceptions are only returned here! There is no callback :()
+    # For example, if there is a bad port number
+    client.execute!.each_with_index {|resp,i|
```
Use do...end for this block, and space after 'resp,'
```ruby
each_with_index do |resp, i|
   ...
end
```

andrewvc replied:
done!

andrewvc commented:
@jordansissel I think I've addressed all your concerns. I also added a new method to dump metadata into a key of your choosing since `@metadata` is effectively broken wrt being able to get data out of it, as @ph and I discovered today. There are tests around this new functionality.

jordansissel commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -35,46 +53,10 @@ class LogStash::Inputs::HTTP_Poller < LogStash::Inputs::Base
+  # If you'd like to work with the request/response metadata
+  # Set this value to the name of the field you'd like to store a nested
+  # hash of metadata.
+  config :metadata_target, :validate => :string, :default => nil
```
Should this be `:default => "@metadata"` ?

andrewvc replied:
yep

jordansissel commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -91,77 +73,101 @@ def run(queue)
+
+    # Some exceptions are only returned here! There is no callback,
+    # for example, if there is a bad port number.
+    client.execute!.each_with_index do |resp, i|
```
Feels weird that sometimes an exception is returned sometimes it's raised. Is this what happens? Can we have one behavior instead?

andrewvc replied:
Yeah, I need to dig into this. I tried to unify this earlier today and hit some inconsistencies. This may need to be passed back upstream to Manticore as a bug. It seems that the failure handler misses some exceptions.

andrewvc replied:
I've submitted a PR to Manticore. This works correctly for now, so I think we should make it a TODO. Once there's a new manticore out its unnecessary.

jordansissel replied:
+1

jordansissel commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -91,77 +73,101 @@ def run(queue)
-      event["message"] = decoded.to_hash
-      event["success"] = response.code >= 200 && response.code <= 299
+  def request_async(queue, name, url)
+    @logger.debug? && @logger.debug("Will get url '#{name}' '#{url}")
```
use a context hash:

```ruby
@logger.debug? && @logger.debug("Async get", :name => name, :url => url)
```

jordansissel commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -91,77 +73,101 @@ def run(queue)
+    started = Time.now
+    client.async.get(url).
+      on_success {|response| handle_success(queue, name, url, response, Time.now - started)}.
+      on_failure {|exception| handle_failure(queue, name, url, exception)}
```
Do you want to pass the duration `Time.now - started` to the on_failure handler also?

jordansissel commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -91,77 +73,101 @@ def run(queue)
+    event.tag("_http_request_failure")
+
+    # This is also in the metadata, but we send it anyone because we want this
+    # persisted by default, whereas metadata isn't. People don't like mysterious errors
```
:+1: mystery reduction! :)

jordansissel commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -91,77 +73,101 @@ def run(queue)
+        "url" => url
+      }
+
+    m["runtime_millis"] = (execution_time*1000).to_i if execution_time
```
My personal preference is to store values as the base unit (bytes, not megabytes; seconds, not milliseconds; meters, not kilometers; etc...)

Any gain in storing millisecond integer value instead of a floating point second value?

andrewvc replied:
Just my java brain showing. I'm just way more used to seeing timestamps as integers. I don't mind making it the float that the `Time` class uses.

jordansissel commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -10,18 +10,36 @@
+# }
+#
+# # This plugin uses metadata, which by default is not serialized
+# # You'll need a filter like this to preserve the poller metadata
```
With the new `metadata_target` setting you've added, this example may not be necessary, but maybe we should keep the example that shows the content of the metadata?

andrewvc replied:
Ah, good call. In fact that example is broken since it adds the metadata as stringified JSON. It's not actually possible to use the metadata due to a bug in `mutate` that makes it incapable of properly copying or renaming metadata.

jordansissel replied:
So should we remove the example?

jordansissel commented:
Completed another pass at reviewing. Tests look OK and they pass for me (woo!). Back to you @andrewvc - I left a few additional comments. Things are looking good though! :)

andrewvc commented:
@jordansissel made the requested changes. How's it looking?

jordansissel commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -91,77 +73,107 @@ def run(queue)
+      request_async(queue, name, url)
+    end
+
+    # TODO: Remove this once our patch to manticore is accepted. The real callback should work
```
can you include a link to the manticore issue/pr in this comment?

jordansissel commented on lib/logstash/inputs/http_poller.rb:
```diff
@@ -91,77 +73,107 @@ def run(queue)
+        name = @urls.keys[i]
+        url = @urls[name]
+        # We can't report the time here because this is as slow as the slowest request
+        # This is all temporary code anyway
```
temporary until the manticore issue is resolved?

andrewvc replied:
Yeah, once that's in we just delete this all.

jordansissel commented:
Code LGTM. Specs pass.

Can you squash this into one (or a smaller number, if 1 is not meaningful) of commits w/ clear commit messages? Some things wont' need to survive merging (like commits about fixing things commented-on within the review process)

andrewvc commented:
@jordansissel I think I'll just make this one big commit if you don't mind. At this point, so much is in flux, there's not much to be gained by breaking it up.

jordansissel commented:
<3

suyograo commented:
nice one! @andrewvc
